### PR TITLE
Speed up loading the initial data, by using the 'value' callback.

### LINKED ISSUE #212: firebase-query is very slow when loading an array with more than just a few items
### Description

I am loading a firebase array with around 12000 items, using a Polymer element like:

```html
<link rel="import" href="/bower_components/polymer/polymer.html">

<link rel="import" href="/bower_components/polymerfire/firebase-query.html">

<dom-module id="loader">
  <template>
    <firebase-query id="query"
                    path="/data"
                    data="{{data}}">
    </firebase-query>
  </template>

  <script>
    (function () {
      'use strict';

      Polymer ({
        is: 'loader',

        properties: {
          data: {
            type: Object,
            value: [],
            notify: true
          }
        },

        ready: function () {
          var self = this;

          setTimeout (function () {
            console.log ('Data load took ' + (self.$.query.__loadEnd - self.$.query.__loadStart) + 'ms');
          }, 60000);
        }
      });
    }) ();
  </script>
</dom-module>
```

The setTimeout () callback is for my instrumentation, which in the current version of firebase-query looks like:

```
--- firebase-query.html.orig	2017-04-28 12:13:59.221771214 +0100
+++ firebase-query.html.instr	2017-04-28 14:31:49.199146378 +0100
@@ -308,6 +308,10 @@
         },
 
         __onFirebaseChildAdded: function(snapshot, previousChildKey) {
+          if (!this.__loadStart) {
+            this.__loadStart = performance.now ();
+          }
+
           var key = snapshot.key;
           var value = snapshot.val();
           var previousChildIndex = this.__indexFromKey(previousChildKey);
@@ -318,6 +322,7 @@
 
           this.__map[key] = value;
           this.splice('data', previousChildIndex + 1, 0, value);
+          this.__loadEnd = performance.now ();
         },
 
         __onFirebaseChildRemoved: function(snapshot) {
```
### Expected outcome

My app should load the data quickly, without interrupting the user experience too severely.

### Actual outcome

My instrumentation timings for several loads are (times in ms):

37324.61
37408.37
36633.60
37129.21
38887.08

Yes, that is nearly 40 seconds to load just 12000 items!  Worse, the browser is completely frozen while this is happening.

### Steps to reproduce

1. Have a decent-sized array in Firebase
2. Load it as above
3. Try to interact with page elements while the load happens

### Browsers Affected

- [x] Chrome
- [ ] Firefox
- [ ] Safari 9
- [ ] Safari 8
- [ ] Safari 7
- [ ] Edge
- [ ] IE 11
- [ ] IE 10

I've tested with Chrome, but I doubt this is browser specific.

### PR DESCRIPTION
**Pull request by dickp**
When 'value' has finished, then start listening for 'child_added'; in
the 'child_added' callback, ignore any additions that we already have.

All polymerfire tests pass.

Fixes #212

### CODE REVIEW

**Comment by dickp** [comment]
I signed it!

File: firebase-query.html [hunk 1]
```diff
@@ -280,18 +280,31 @@
...
             this.syncToMemory(function() {
+              this.__map = {};
               this.set('data', this.zeroValue);
             });
           }
 
           if (query) {
-            query.on('child_added', this.__onFirebaseChildAdded, this.__onError, this);
+            /* The 'value' event doesn't order the data, so we have to
+             * revert to the slow path if ordering is requested
```
Comment (mikelehen) [inline firebase-query.html]: The DataSnapshot provided to the 'value' event is in fact ordered.  But .val() returns it as a plain JS Object, which is not ordered.  To preserve ordering, you can iterate the DataSnapshot using the .forEach() method.

File: firebase-query.html [hunk 2]
```diff
@@ -309,17 +322,45 @@
...
+
+              this.__map[key] = value;
+              data.push(value);
+            }
+
+            this.set('data', data);
+          }
+
+          /* Now start listening for changes */
+          this.query.on('child_added', this.__onFirebaseChildAdded, this.__onError, this);
```
Comment (mikelehen) [inline firebase-query.html]: You can attach this right away but ignore the events you get before the first 'value' event.

Comment (tjmonsi) [inline firebase-query.html]: @mikelehen I forked @dickp 's branch and asked for a PR on his side that is connected to here, using what you have suggested

https://github.com/dickp/polymerfire/pull/1

**Comment by dickp** [comment]
I signed it! Again!

**Comment by tjmonsi** [comment]
@dickp I think it is @ecgtheow that is not signed in?

**Comment by dickp** [comment]
@tjmonsi Someone in the CLA office didn't think I was authorised to sign on behalf of my company, so I tried again.

When I get around to it, I'll be converting @ecgtheow to an organisation account anyway.

**Comment by mikelehen** [comment]
Changes look good to me (thanks)!

**Comment by justweb1** [comment]
I think @ecgtheow is the one that needs to sign the CLA.

**Comment by dickp** [comment]
@justweb1 I have been trying to sign the CLA as @ecgtheow, but someone in the CLA office keeps deciding I don't have the authority and voiding the signature.  I've pointed them at the UK government's companies registration site, but it's like talking to a brick wall.

I'll try signing it yet again now.

**Comment by dickp** [comment]
Aaaand they 'voided' it again. I give up.

**Comment by mbleigh** [comment]
@dickp would you be willing to re-open this in a new PR as a single commit from just your personal account? That should make the CLA bot happy (I hope).

**Comment by dickp** [comment]
I don't think it's the CLA bot.  Every time I try to sign it, I get an email back that says:

"Google Contributor License Agreement for Ecgtheow Ltd has been voided for the following reason:

Please have this document signed by an authorized executive. ..."

No "please show us some documentation" (I tried to), not that that would help as emails to cla-submissions@google.com seem to go straight to /dev/null anyway.

It's a bank holiday weekend here and I'm making it a long one.  I might have some time next week to look at this.

**Comment by justweb1** [comment]
@dickp if you squash all if the changes into one commit using your username instead if the company you will be able to submit a new OR with no issue.

**Comment by christophe-g** [comment]
Hey, 
That seems to be a great PR - which should solve some of the perf issue I have with firebase-query. 
Any news re CLA or new PR?
thanks!

**Comment by mbleigh** [comment]
Supplanted by new PR with functioning CLA.

**Pull request closed without merging**